marlowe/test/test-report.md

```diff
@@ -0,0 +1,543 @@
+- Informal on- and off-chain tests use Jupyter notebooks.
+- Code coverage was checked using the GHC [HPC](https://wiki.haskell.org/Haskell_program_coverage) capability.
+
+
+## Summary of Testing History and Results
+
+The Marlowe validator implementation currently passes all property-based and on-chain tests. The testing code and process outlined above revealed several issues that were further scrutinized:
+1. The test for ["Constraint 4: No output to script on close"](../specification/ReadMe.md#constraint-4-no-output-to-script-on-close) of the Marlowe-Cardano specification originally failed. This failure was eliminated with a minor modification to the semantics validator.
+2. Testing revealed an ambiguity regarding how a payment of no value should be warned as `ReduceNonPositivePay` versus `ReducePartialPay`. In documentation we clarify that an intentional payment of zero results in the warning `ReduceNonPositivePay`, which means "the contract was meant to pay zero or less but that is impossible so it pays nothing", whereas a payment that happens to be zero because of insufficient funds results in the warning `ReducePartialPay`, which means "the contract is meant to pay more than there is, so it pays what there is".
+3. The `computeTransaction` function does not guard for the preconditions (1) that internal account balances are positive and (2) that duplicate `(Account, Token)` pairs are not present in the association map `accounts :: PlutusTx.Map (Account, Token) Integer`.
+    1. Marlowe's semantics validator *guards* against the first case by failing to execute any transaction that begins with a non-positive balance in an internal account. *This causes permanent locking of funds in the contract.*
+    2. The semantics validator *does not guard* against duplicate accounts, in which circumstance an invalid initial state would be provided to `computeTransaction`, resulting in behavior that violates the semantics specification. *This potentially causes semantically invalid behavior of the contract.*
+4. A role-based payments to a party with a blank token name in a contract where no roles currency was defined (effectively making Ada the roles currency) can be withdrawn from the payout validator by anyone at all.
+5. Slot number to POSIX millisecond time conversion from the node to the Plutus script context necessitates meticulous attention to time values in Marlowe contracts. In some cases, sub-second mismatches will cause a validation failure, but that can be mitigated by resubmitting the transaction with the correct slot-number validity interval.
+6. The POSIX-millisecond validity intervals in the Plutus script context may be open, closed, or half-open/half-closed, but the slot-number validity intervals at the node are always closed on the left and open on the right. This situation has a potential (similar to Issue 5 above) to make the temporal aspect of validation finicky.
```

Comment:
> The semantics validator *does not guard* against duplicate accounts, in which circumstance an invalid initial state would be provided to `computeTransaction`, resulting in behavior that violates the semantics specification. *This potentially causes semantically invalid behavior of the contract.*

@palas, this has bothered me because we do guard against non-positive account balances as a precondition. We could add a `any . foldMap ... accounts` guard against this, but `foldMap` on an association list is somewhat expensive, plus we'd should probably do this for `choices` and `boundValues`, too.